File: app/src/main/java/com/example/wearablecollector/SensorDataRepository.kt
package com.example.wearablecollector

import androidx.lifecycle.MutableLiveData

object SensorDataRepository {
    val heartRate = MutableLiveData<Int>(0)
    val hrv = MutableLiveData<Int>(0) // Placeholder, as standard HR service might not give explicit HRV
    val gsr = MutableLiveData<Float>(0f)
    val eda = MutableLiveData<Float>(0f)
    val stressLevel = MutableLiveData<Int>(0)
    val status = MutableLiveData<String>("Disconnected")
    val scannedDevices = MutableLiveData<List<android.bluetooth.BluetoothDevice>>(emptyList())

    fun addScannedDevice(device: android.bluetooth.BluetoothDevice) {
        val currentList = scannedDevices.value.orEmpty().toMutableList()
        if (currentList.none { it.address == device.address }) {
            currentList.add(device)
            scannedDevices.postValue(currentList)
        }
    }
    
    fun clearScannedDevices() {
        scannedDevices.postValue(emptyList())
    }

    fun updateHeartRate(hr: Int) {
        heartRate.postValue(hr)
    }

    fun updateHrv(rrInterval: Int) {
        // Calculate HRV and Stress using the new logic
        val result = com.example.wearablecollector.logic.StressCalculator.processRR(rrInterval)
        hrv.postValue(result.hrv)
        stressLevel.postValue(result.stressLevel)
    }

    fun updateGsr(value: Float) {
        gsr.postValue(value)
    }
    
    fun updateEda(value: Float) {
        eda.postValue(value)
    }

    fun updateStatus(newStatus: String) {
        status.postValue(newStatus)
        if (newStatus == "Disconnected") {
            com.example.wearablecollector.logic.StressCalculator.reset()
            heartRate.postValue(0)
            hrv.postValue(0)
            stressLevel.postValue(0)
        }
    }
}

File: ios/Vibree/Vibree/ViewModels/SensorDataRepository.swift
import Foundation
import Combine
import CoreBluetooth

class SensorDataRepository: ObservableObject {
    static let shared = SensorDataRepository()
    
    @Published var heartRate: Int = 0
    @Published var hrv: Int = 0
    @Published var gsr: Float = 0.0
    @Published var eda: Float = 0.0
    @Published var stressLevel: Int = 0
    @Published var status: String = "Disconnected"
    @Published var scannedDevices: [CBPeripheral] = []
    
    private let stressCalculator = StressCalculator()
    
    private init() {}
    
    func addScannedDevice(_ device: CBPeripheral) {
        if !scannedDevices.contains(where: { $0.identifier == device.identifier }) {
            scannedDevices.append(device)
        }
    }
    
    func clearScannedDevices() {
        scannedDevices.removeAll()
    }
    
    func updateHeartRate(_ hr: Int) {
        self.heartRate = hr
    }
    
    func updateHrv(rrInterval: Int) {
        let result = stressCalculator.processRR(rrMs: rrInterval)
        
        self.hrv = result.hrv
        self.stressLevel = result.stressLevel
        
        // In a real app, save to CoreData or Realm here
        if result.hrv > 0 {
            print("Logged: HR: \(heartRate), HRV: \(result.hrv), Stress: \(result.stressLevel)")
        }
    }
    
    func updateGsr(_ value: Float) {
        self.gsr = value
    }
    
    func updateEda(_ value: Float) {
        self.eda = value
    }
    
    func updateStatus(_ newStatus: String) {
        self.status = newStatus
        if newStatus == "Disconnected" {
            stressCalculator.reset()
            self.heartRate = 0
            self.hrv = 0
            self.stressLevel = 0
        }
    }
}
